os/drivers/error_report/error_report_drv.c

```diff
@@ -0,0 +1,301 @@
+				INSERT_INFINITE_WAITQ(pos, tcb, cpu);
+			}
+		} 
+		else if (!g_thd_waitq[pos].pid) {
+			/* Add new entry */
+			INSERT_INFINITE_WAITQ(pos, tcb, cpu);
+		}
+		else {
+			/* Collision resolution, find the next empty slot */
+			int i;
+			for (i = pos+1; i != pos; i = (i+1)%CONFIG_MAX_TASKS) {
+				if (g_thd_waitq[i].pid == 0) {
+					/* Found an empty slot, insert values here */
+					INSERT_INFINITE_WAITQ(i, tcb, cpu);
+					break;
+				}
+			}
+			if (i == pos) {
+				nwerrdbg("Cannot find a slot to insert values\n");
+			}
+		}		
+	}
+	else {
+		/* Maybe the pid was blocked during the 
+		 * last sampling instance, and 
+		 * is active again. Reset the data 
+		 * in this case. 
+		 */
+		/* Hash pid over the waitq range */
+		pos = tcb->pid % CONFIG_MAX_TASKS;
+		if ((g_thd_waitq[pos].pid == tcb->pid)) {
+			/* Reset the count, indicating that is free for insert */
+			DELETE_INFINITE_WAITQ(pos);
+		} else if (g_thd_waitq[pos].pid) {
+			/* Collision resolution, find the slot using linear search */
+			int i;
+			for (i = pos+1; i != pos; i = (i+1)%CONFIG_MAX_TASKS) {
+				if (g_thd_waitq[i].pid == tcb->pid) {
+					/* Found an empty slot, insert values here */
+					DELETE_INFINITE_WAITQ(i);
+					break;
+				}
+			}
+		}
+	}
+}
+
+
+
+/************************************************************************************
+ * Name: error_report_ioctl
+ *
+ * Description:  The ioctl method for error reporting.
+ *
+ ************************************************************************************/
+static int error_report_ioctl(FAR struct file *filep, int cmd, unsigned long arg)
+{
+	int ret = -EINVAL;
+	
+
+	nwerr_vdbg("cmd: %d arg: %ld\n", cmd, arg);
+
+	/* Handle built-in ioctl commands */
+
+	switch (cmd) {
+	case ERIOC_GET_ENTRY:
+	{
+		thread_entry_t *thd = (thread_entry_t *) ((uintptr_t)arg);
+		thd->entry = prv_fetch_taskaddr(thd->pid);
+		ret = OK;
+	}
+	break;
+	case ERIOC_CHECK_INFWAIT:
+		sched_foreach(prv_tcb_handler, NULL);
+		ret = OK;
+	break;
+	case ERIOC_GET_BACKTRACE:
+	{
+		infinty_waitdata_t *waitdata = (infinty_waitdata_t *) ((uintptr_t)arg);
+		struct tcb_s *tcb = sched_gettcb(waitdata->pid);
+		if (tcb == NULL) {
+			break;
+		}
+		waitdata->ncalls = 0;
+		/* Read the functions from the callstack into waitdata structure */
+		uint32_t *ptr;
+		ptr = (uint32_t *)tcb->xcp.regs[13];
+		while ((void *)ptr < tcb->adj_stack_ptr && (waitdata->ncalls < CONFIG_ERROR_REPORT_BACKTRACE_MAX_DEPTH)) {
+			if (*ptr >= 0x04000000 && *ptr <= 0x04800000) {
+				waitdata->backtrace[waitdata->ncalls] = *ptr - 4;
+				nwerr_vdbg("Stack: 0x%08x\n", *ptr - 4);
+				waitdata->ncalls++;
+			}
+			ptr++;
+		}
+		/* Fill out the remaining slots in waitdata with 0 */
+		while (waitdata->ncalls < CONFIG_ERROR_REPORT_BACKTRACE_MAX_DEPTH) {
+			waitdata->backtrace[waitdata->ncalls] = 0;
+			waitdata->ncalls++;
+		}
+		waitdata->task_state = tcb->task_state;
+		waitdata->entry = prv_fetch_taskaddr(waitdata->pid);
```

Comment:
sched_gettcb is called twice in this case. Line 252 and here.

Reply:
Ok, I factored the `prv_fetch_taskaddr` to include a macro `RETURN_ADDR_FROM_TCB` for fetching task addr given a tcb pointer. So, the above statement can be replaced by a macro call to `RETURN_ADDR_FROM_TCB`.